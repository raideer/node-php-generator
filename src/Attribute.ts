export type AttributeArgument = string | number | boolean | null;
export type AttributeArguments = Record<string, AttributeArgument>;

export class Attribute {
  private name: string;
  private args: AttributeArguments;

  constructor(name: string, args: AttributeArguments = {}) {
    this.name = name;
    this.args = args;
  }

  public getName(): string {
    return this.name;
  }

  public getArguments(): AttributeArguments {
    return this.args;
  }
}
