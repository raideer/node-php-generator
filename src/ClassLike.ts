import { withAttributes } from "@/mixins/withAttributes";
import { withComments } from "@/mixins/withComments";
import { PhpNamespace } from "@/PhpNamespace";

export enum ClassVisibility {
  PUBLIC = "public",
  PROTECTED = "protected",
  PRIVATE = "private",
}

export class ClassLike extends withComments(withAttributes(class {})) {
  private namespace: PhpNamespace | null = null;

  private name: string | undefined;

  constructor(name: string, namespace: PhpNamespace | null = null) {
    super();

    this.name = name;
    this.namespace = namespace;
  }

  getName(): string | undefined {
    return this.name;
  }

  setName(name: string | undefined): this {
    this.name = name;

    return this;
  }

  isClass(): boolean {
    return false;
  }

  isTrait(): boolean {
    return false;
  }

  isInterface(): boolean {
    return false;
  }

  isEnum(): boolean {
    return false;
  }
}
