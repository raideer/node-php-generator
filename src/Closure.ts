import { asFunction } from "./mixins/asFunction";
import { withAttributes } from "./mixins/withAttributes";
import { Parameter } from "./Parameter";

export class Closure extends asFunction(withAttributes(class {})) {
  private uses: Parameter[] = [];

  getUses(): Parameter[] {
    return this.uses;
  }

  setUses(uses: Parameter[]): this {
    this.uses = uses;

    return this;
  }

  addUse(name: string): this {
    this.uses.push((new Parameter()).setName(name));

    return this;
  }
}
