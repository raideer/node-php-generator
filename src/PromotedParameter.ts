import { CanValidate } from './Api/CanValidate';
import { HasVisibility } from './Api/HasVisibility';
import { Parameter } from './Parameter';
import { Visibility } from './types';

export class PromotedParameter
  extends Parameter
  implements HasVisibility, CanValidate
{
  private visibility: Visibility | undefined;
  private readOnly: boolean = false;

  public getVisibility(): Visibility | undefined {
    return this.visibility;
  }

  public setVisibility(visibility: Visibility | undefined): this {
    this.visibility = visibility;
    return this;
  }

  public setPublic(): this {
    this.visibility = 'public';
    return this;
  }

  public isPublic(): boolean {
    return this.visibility === 'public';
  }

  public setPrivate(): this {
    this.visibility = 'private';
    return this;
  }

  public isPrivate(): boolean {
    return this.visibility === 'private';
  }

  public setProtected(): this {
    this.visibility = 'protected';
    return this;
  }

  public isProtected(): boolean {
    return this.visibility === 'protected';
  }

  public setReadOnly(value: boolean = true): this {
    this.readOnly = value;
    return this;
  }

  public isReadOnly(): boolean {
    return this.readOnly;
  }

  public validate(): void {
    if (this.readOnly && !this.getType()) {
      throw new Error(
        'Promoter parameter cannot be read-only if it does not have a type'
      );
    }
  }

  public toString(): string {
    //TODO: Implement
    return '';
  }
}
