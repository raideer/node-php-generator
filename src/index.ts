export { Attribute } from './Attribute';
export { ClassType } from './ClassType';
export { Closure } from './Closure';
export { Constant } from './Constant';
export { EnumCase } from './EnumCase';
export { EnumType } from './EnumType';
export { GlobalFunction } from './GlobalFunction';
export { InterfaceType } from './InterfaceType';
export { Literal } from './Literal';
export { Method } from './Method';
export { Parameter } from './Parameter';
export { PhpFile } from './PhpFile';
export { PhpNamespace } from './PhpNamespace';
export { Printer } from './Printer';
export { PromotedParameter } from './PromotedParameter';
export { Property } from './Property';
export { PsrPrinter } from './PsrPrinter';
export { TraitType } from './TraitType';
export { TraitUse } from './TraitUse';
