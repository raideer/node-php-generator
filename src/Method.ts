import { asFunction } from "mixins/asFunction";
import { withAttributes } from "mixins/withAttributes";
import { withComments } from "mixins/withComments";
import { withName } from "mixins/withName";
import { withVisibility } from "mixins/withVisibility";
import { PromotedParameter } from "./PromotedParameter";

export class Method extends asFunction(withVisibility(withAttributes(withComments(withName(class {}))))) {
  public static: boolean = false;
  public final: boolean = false;
  public abstract: boolean = false;

  setStatic(isStatic: boolean): this {
    this.static = isStatic;

    return this;
  }

  isStatic(): boolean {
    return this.static;
  }

  setFinal(final: boolean): this {
    this.final = final;

    return this;
  }

  isFinal(): boolean {
    return this.final;
  }

  setAbstract(abstract: boolean): this {
    this.abstract = abstract;

    return this;
  }

  isAbstract(): boolean {
    return this.abstract;
  }

  addPromotedParameter(name: string, defaultValue: string | null | undefined): this {
    const parameter = new PromotedParameter();
    parameter.setName(name);

    if (defaultValue !== undefined) {
      parameter.setDefaultValue(defaultValue);
    }

    this.parameters[name] = parameter;

    return this;
  }
}
