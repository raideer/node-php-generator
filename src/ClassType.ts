import { ClassLike } from "@/ClassLike";
import { PhpNamespace } from "@/PhpNamespace";
import { Constant } from "@/Constant";
import { Method } from "@/Method";
import { Property } from "@/Property";
import { withConstants } from "@/mixins/withConstants";
import { withMethods } from "@/mixins/withMethods";
import { withProperties } from "@/mixins/withProperties";
import { withTraits } from "@/mixins/withTraits";
import { TraitUse } from "@/TraitUse";

export type ClassTypeMember = Constant | Method | Property | TraitUse;

export class ClassType extends withConstants(withMethods(withProperties(withTraits(ClassLike)))) {
  private final = false;
  private abstract = false;
  private extends: string | null = null;
  private readOnly = false;
  private implements: string[] = [];

  isClass(): boolean {
    return true;
  }

  constructor(name: string, namespace: PhpNamespace | null = null) {
    super(name, namespace);
  }

  setFinal(final: boolean = true): this {
    this.final = final;

    return this;
  }

  isFinal(): boolean {
    return this.final;
  }

  setAbstract(abstract: boolean = true): this {
    this.abstract = abstract;

    return this;
  }

  setReadOnly(readOnly: boolean = true): this {
    this.readOnly = readOnly;

    return this;
  }

  isReadOnly(): boolean {
    return this.readOnly;
  }

  isAbstract(): boolean {
    return this.abstract;
  }

  setExtends(ext: string | null): this {
    this.extends = ext;

    return this;
  }

  getExtends(): string | null {
    return this.extends;
  }

  setImplements(impl: string[]): this {
    this.implements = impl;

    return this;
  }

  getImplements(): string[] {
    return this.implements;
  }

  addMember(member: ClassTypeMember, override: boolean = false): this {
    const name = member.getName() as string;

    if (member instanceof Constant) {
      this.constants[name] = member;
    } else if (member instanceof Method) {
      this.methods[name] = member;
    } else if (member instanceof Property) {
      this.properties[name] = member;
    } else if (member instanceof TraitUse) {
      this.traits[name] = member;
    }

    return this;
  }
}
