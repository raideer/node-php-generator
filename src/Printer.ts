import { GlobalFunction } from './GlobalFunction';
import { Closure } from './Closure';
import { Method } from './Method';
import { InterfaceType } from './InterfaceType';
import { ClassType } from './ClassType';
import { TraitType } from './TraitType';
import { EnumType } from './EnumType';
import { PhpNamespace } from './PhpNamespace';
import { PhpFile } from './PhpFile';
import { Dumper } from './Dumper';
import { Helpers } from './Helpers';
import { Attribute } from './Attribute';
import { PromotedParameter } from './PromotedParameter';
import { Constant } from './Constant';
import { Property } from './Property';

export class Printer {
  public wrapLength: number = 120;
  public indentation: string = '    ';
  public linesBetweenProperties: number = 0;
  public linesBetweenMethods: number = 2;
  public linesBetweenUseTypes: number = 0;
  public returnTypeColon: string = ': ';
  public bracesOnNextLine: boolean = true;
  public singleParameterOnOneLine: boolean = false;
  public omitEmptyNamespaces: boolean = true;
  protected namespace: PhpNamespace | null = null;
  protected dumper: Dumper;
  private resolveTypes: boolean = true;

  constructor() {
    this.dumper = new Dumper();
  }

  public printFunction(
    func: GlobalFunction,
    namespace: PhpNamespace | null = null
  ): string {
    this.namespace = this.resolveTypes ? namespace : null;
    const line =
      'function ' + (func.getReturnReference() ? '&' : '') + func.getName();
    const returnType = this.printReturnType(func);
    const params = this.printParameters(
      func,
      line.length + returnType.length + 2
    );
    const body = Helpers.simplifyTaggedNames(func.getBody(), this.namespace);
    const braceOnNextLine = this.isBraceOnNextLine(
      params.includes('\n'),
      !!returnType
    );

    return (
      this.printDocComment(func) +
      this.printAttributes(func.getAttributes()) +
      line +
      params +
      returnType +
      (braceOnNextLine ? '\n' : ' ') +
      '{\n' +
      this.indent(body) +
      '}\n'
    );
  }

  public printClosure(
    closure: Closure,
    namespace: PhpNamespace | null = null
  ): string {
    this.namespace = this.resolveTypes ? namespace : null;
    const uses = closure
      .getUses()
      .map((param) => (param.isReference() ? '&' : '') + '$' + param.getName());
    const useStr =
      uses.length > 1 && uses.join(', ').length > this.wrapLength
        ? '\n' + this.indentation + uses.join(',\n' + this.indentation) + ',\n'
        : uses.join(', ');
    const body = Helpers.simplifyTaggedNames(closure.getBody(), this.namespace);

    return (
      this.printAttributes(closure.getAttributes(), true) +
      'function ' +
      (closure.getReturnReference() ? '&' : '') +
      this.printParameters(closure) +
      (uses.length ? ' use (' + useStr + ')' : '') +
      this.printReturnType(closure) +
      ' {\n' +
      this.indent(body) +
      '}'
    );
  }

  public printMethod(
    method: Method,
    namespace: PhpNamespace | null = null,
    isInterface: boolean = false
  ): string {
    this.namespace = this.resolveTypes ? namespace : null;
    method.validate();
    const line =
      (method.isAbstract() && !isInterface ? 'abstract ' : '') +
      (method.isFinal() ? 'final ' : '') +
      (method.getVisibility() ? method.getVisibility() + ' ' : '') +
      (method.isStatic() ? 'static ' : '') +
      'function ' +
      (method.getReturnReference() ? '&' : '') +
      method.getName();
    const returnType = this.printReturnType(method);
    const params = this.printParameters(
      method,
      line.length + returnType.length + this.indentation.length + 2
    );
    let body = Helpers.simplifyTaggedNames(method.getBody(), this.namespace);
    body = Helpers.normalize(body).trimStart();
    body = body.trimEnd() + '\n';
    const braceOnNextLine = this.isBraceOnNextLine(
      params.includes('\n'),
      !!returnType
    );

    return (
      this.printDocComment(method) +
      this.printAttributes(method.getAttributes()) +
      line +
      params +
      returnType +
      (method.isAbstract() || isInterface
        ? ';\n'
        : (braceOnNextLine ? '\n' : ' ') + '{\n' + this.indent(body) + '}\n')
    );
  }

  public printClass(
    classType: ClassType | InterfaceType | TraitType | EnumType,
    namespace: PhpNamespace | null = null
  ): string {
    this.namespace = this.resolveTypes ? namespace : null;
    classType.validate();
    const resolver = this.namespace
      ? (s: string) => this.namespace!.simplifyType(s)
      : (s: string) => s;

    const traits =
      classType instanceof ClassType ||
      classType instanceof TraitType ||
      classType instanceof EnumType
        ? classType
            .getTraits()
            .map((trait) => {
              const resolutions = Helpers.simplifyTaggedNames(
                trait.getResolutions().join(';\n'),
                this.namespace
              );
              return (
                this.printDocComment(trait) +
                'use ' +
                resolver(trait.getName()) +
                (resolutions
                  ? ' {\n' + this.indent(resolutions) + ';\n}\n'
                  : ';\n')
              );
            })
            .join('')
        : '';

    const cases =
      classType instanceof EnumType
        ? classType
            .getCases()
            .map((enumCase) => {
              return (
                this.printDocComment(enumCase) +
                this.printAttributes(enumCase.getAttributes()) +
                'case ' +
                enumCase.getName() +
                (enumCase.getValue() === null
                  ? ''
                  : ' = ' + this.dump(enumCase.getValue())) +
                ';\n'
              );
            })
            .join('')
        : '';

    const readOnlyClass = classType instanceof ClassType;
    const consts = classType
      .getConstants()
      .map((constant) => this.printConstant(constant))
      .join('');
    const methods = classType
      .getMethods()
      .map((method) => {
        if (readOnlyClass && method.getName() === 'constructor') {
          const clonedMethod = method.clone();

          clonedMethod.getParameters().forEach((param) => {
            if (param instanceof PromotedParameter) {
              param.setReadOnly(false);
            }
          });

          return this.printMethod(
            clonedMethod,
            namespace,
            classType.isInterface()
          );
        }

        return this.printMethod(method, namespace, classType.isInterface());
      })
      .join('');
    const properties =
      classType instanceof ClassType || classType instanceof TraitType
        ? classType
            .getProperties()
            .map((property) => this.printProperty(property, readOnlyClass))
            .join('')
        : '';

    const members = [traits, consts, cases, properties, methods]
      .filter(Boolean)
      .join('\n');

    let classExtends: string[] = [];

    if (classType instanceof ClassType || classType instanceof InterfaceType) {
      if (classType.getExtends()) {
        if (classType instanceof ClassType) {
          classExtends = [classType.getExtends()!];
        } else {
          classExtends = classType.getExtends();
        }
      }
    }

    const line = [
      classType instanceof ClassType && classType.isAbstract()
        ? 'abstract'
        : null,
      classType instanceof ClassType && classType.isFinal() ? 'final' : null,
      classType instanceof ClassType && classType.isReadOnly()
        ? 'readonly'
        : null,
      classType instanceof ClassType ? 'class ' + classType.getName() : null,
      classType instanceof InterfaceType
        ? 'interface ' + classType.getName()
        : null,
      classType instanceof TraitType ? 'trait ' + classType.getName() : null,
      classType instanceof EnumType
        ? 'enum ' +
          classType.getName() +
          (classType.getType()
            ? this.returnTypeColon + classType.getType()
            : '')
        : null,
      classExtends.length
        ? 'extends ' + classExtends.map(resolver).join(', ')
        : null,
      (classType instanceof ClassType || classType instanceof EnumType) &&
      classType.getImplements().length
        ? 'implements ' + classType.getImplements().map(resolver).join(', ')
        : null,
    ]
      .filter(Boolean)
      .join(' ');

    return (
      this.printDocComment(classType) +
      this.printAttributes(classType.getAttributes()) +
      line +
      '\n{\n' +
      this.indent(members) +
      '}\n'
    );
  }

  public printEnum(enumType: EnumType): string {
    return this.printClass(enumType);
  }

  public printTrait(traitType: TraitType): string {
    return this.printClass(traitType);
  }

  public printInterface(interfaceType: InterfaceType): string {
    return this.printClass(interfaceType);
  }

  public printNamespace(namespace: PhpNamespace): string {
    this.namespace = this.resolveTypes ? namespace : null;
    const name = namespace.getName();
    const uses = [
      this.printUses(namespace),
      this.printUses(namespace, PhpNamespace.NameFunction),
      this.printUses(namespace, PhpNamespace.NameConstant),
    ]
      .filter(Boolean)
      .join('\n'.repeat(this.linesBetweenUseTypes));

    const items = [
      ...namespace.getClasses().map((cls) => this.printClass(cls, namespace)),
      ...namespace
        .getFunctions()
        .map((fn) => this.printFunction(fn, namespace)),
    ].join('\n');

    const body = (uses ? uses + '\n' : '') + items;

    return namespace.hasBracketedSyntax()
      ? `namespace${name ? ` ${name}` : ''} {\n${this.indent(body)}}\n`
      : `${name ? `namespace ${name};\n\n` : ''}${body}`;
  }

  public printFile(file: PhpFile): string {
    const namespaces = file
      .getNamespaces()
      .map((ns) => {
        if (
          !this.omitEmptyNamespaces ||
          ns.getClasses().length ||
          ns.getFunctions().length
        ) {
          return this.printNamespace(ns);
        }
        return '';
      })
      .filter(Boolean)
      .join('\n\n');

    return (
      `<?php\n` +
      (file.getComment() ? '\n' + this.printDocComment(file) : '') +
      '\n' +
      (file.hasStrictTypes() ? 'declare(strict_types=1);\n\n' : '') +
      namespaces
    );
  }

  protected printUses(
    namespace: PhpNamespace,
    of: string = PhpNamespace.NameNormal
  ): string {
    const prefix = {
      [PhpNamespace.NameNormal]: '',
      [PhpNamespace.NameFunction]: 'function ',
      [PhpNamespace.NameConstant]: 'const ',
    }[of];
    return Object.entries(namespace.getUses(of))
      .map(([alias, original]) => {
        return Helpers.extractShortName(original) === alias
          ? `use ${prefix}${original};\n`
          : `use ${prefix}${original} as ${alias};\n`;
      })
      .join('');
  }

  protected printParameters(
    func: GlobalFunction | Method | Closure,
    column: number = 0
  ): string {
    const special = func
      .getParameters()
      .some(
        (param) =>
          param instanceof PromotedParameter ||
          param.getAttributes().length ||
          param.getComment()
      );

    if (
      !special ||
      (this.singleParameterOnOneLine && func.getParameters().length === 1)
    ) {
      const line = this.formatParameters(func, false);
      if (!line.includes('\n') && line.length + column <= this.wrapLength) {
        return line;
      }
    }
    return this.formatParameters(func, true);
  }

  private formatParameters(
    func: GlobalFunction | Method | Closure,
    multiline: boolean
  ): string {
    const params = func.getParameters();

    const res = params
      .map((param) => {
        const variadic =
          func.isVariadic() && param === params[params.length - 1];
        const attrs = this.printAttributes(param.getAttributes(), true);

        return (
          this.printDocComment(param) +
          (attrs ? (multiline ? attrs.slice(0, -1) + '\n' : attrs) : '') +
          (param instanceof PromotedParameter
            ? (param.getVisibility() || 'public') +
              (param.isReadOnly() && param.getType() ? ' readonly' : '') +
              ''
            : '') +
          this.printType(param.getType(), param.isNullable()) +
          ' ' +
          (param.isReference() ? '&' : '') +
          (variadic ? '...' : '') +
          '$' +
          param.getName() +
          (param.getDefaultValue() && !variadic
            ? ' = ' + this.dump(param.getDefaultValue())
            : '') +
          (multiline ? ',\n' : ', ')
        );
      })
      .join('');

    return multiline
      ? '(\n' + this.indent(res) + ')'
      : '(' + res.slice(0, -2) + ')';
  }

  private printConstant(constant: Constant): string {
    const def =
      (constant.isFinal() ? 'final ' : '') +
      (constant.getVisibility() ? constant.getVisibility() + ' ' : '') +
      'const ' +
      this.printType(constant.getType(), false) +
      ' ' +
      constant.getName() +
      ' = ';

    return (
      this.printDocComment(constant) +
      this.printAttributes(constant.getAttributes()) +
      def +
      this.dump(constant.getValue()) +
      ';\n'
    );
  }

  private printProperty(
    property: Property,
    readOnlyClass: boolean = false
  ): string {
    property.validate();
    const type = property.getType();
    const def =
      (property.getVisibility() || 'public') +
      (property.isStatic() ? ' static' : '') +
      (!readOnlyClass && property.isReadOnly() && type ? ' readonly' : '') +
      ' ' +
      this.printType(type, property.isNullable()) +
      ' ' +
      '$' +
      property.getName();

    return (
      this.printDocComment(property) +
      this.printAttributes(property.getAttributes()) +
      def +
      (property.getValue() === null && !property.isInitialized()
        ? ''
        : ' = ' + this.dump(property.getValue())) +
      ';\n'
    );
  }

  protected printType(type: string | undefined, nullable: boolean): string {
    if (!type) return '';
    if (this.namespace) {
      type = this.namespace.simplifyType(type);
    }
    return nullable ? '?' + type : type;
  }

  protected printDocComment(commentable: {
    getComment(): string | undefined;
  }): string {
    const multiLine =
      commentable instanceof GlobalFunction ||
      commentable instanceof Method ||
      commentable instanceof ClassType ||
      commentable instanceof PhpFile;
    return Helpers.formatDocComment(commentable.getComment() || '', multiLine);
  }

  protected printReturnType(func: GlobalFunction | Method | Closure): string {
    const type = this.printType(func.getReturnType(), func.isReturnNullable());
    return type ? this.returnTypeColon + type : '';
  }

  protected printAttributes(
    attrs: Attribute[],
    inline: boolean = false
  ): string {
    if (!attrs.length) return '';
    this.dumper.indentation = this.indentation;
    const items = attrs.map((attr) => {
      const args = this.dumper.format('...?:', attr.getArguments());
      return this.printType(attr.getName(), false) + (args ? `(${args})` : '');
    });
    return inline ? `#[${items.join(', ')}] ` : `#[${items.join(']\n#[')}]\n`;
  }

  protected indent(s: string): string {
    s = s.replace(/\t/g, this.indentation);
    return s.replace(/(?:^|[\r\n]+)(?=[^\r\n])/g, '$&' + this.indentation);
  }

  protected dump(varToDump: any, column: number = 0): string {
    this.dumper.indentation = this.indentation;
    this.dumper.wrapLength = this.wrapLength;
    let s = this.dumper.dump(varToDump, column);
    s = Helpers.simplifyTaggedNames(s, this.namespace);
    return s;
  }

  protected isBraceOnNextLine(
    multiLine: boolean,
    hasReturnType: boolean
  ): boolean {
    return this.bracesOnNextLine && (!multiLine || hasReturnType);
  }
}
