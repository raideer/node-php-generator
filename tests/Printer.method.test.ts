import { Method } from '../src/Method';
import { Printer } from '../src/Printer';
import { PromotedParameter } from '../src/PromotedParameter';

test('prints basic method', () => {
  const printer = new Printer();
  const method = new Method('foo');

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints public method', () => {
  const printer = new Printer();
  const method = new Method('foo');
  method.setPublic();

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints private method', () => {
  const printer = new Printer();
  const method = new Method('foo');
  method.setPrivate();

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints protected method', () => {
  const printer = new Printer();
  const method = new Method('foo');
  method.setProtected();

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints static method', () => {
  const printer = new Printer();
  const method = new Method('foo');
  method.setStatic();

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints abstract method', () => {
  const printer = new Printer();
  const method = new Method('foo');
  method.setAbstract();

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints final method', () => {
  const printer = new Printer();
  const method = new Method('foo');
  method.setFinal();

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints method with return type', () => {
  const printer = new Printer();
  const method = new Method('foo');
  method.setReturnType('string');

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints method with nullable return type', () => {
  const printer = new Printer();
  const method = new Method('foo');
  method.setReturnType('string');
  method.setReturnNullable(true);

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints method with parameters', () => {
  const printer = new Printer();
  const method = new Method('foo');
  method.addParameter('param1', 'default1');
  method.addParameter('param2', 'default2');

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints method with typed parameters', () => {
  const printer = new Printer();
  const method = new Method('foo');
  const param = method.addParameter('param');
  param.setType('string');
  param.setNullable(true);

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints method with promoted parameters', () => {
  const printer = new Printer();
  const method = new Method('foo');
  method.addPromoterParameter('param', 'default');

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints method with attributes', () => {
  const printer = new Printer();
  const method = new Method('foo');
  method.addAttribute('Route', { path: '/foo' });

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints method with doc comment', () => {
  const printer = new Printer();
  const method = new Method('foo');
  method.addComment('This is a method');
  method.addComment('@param string $param Description');
  method.addComment('@return void');

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints method with reference return', () => {
  const printer = new Printer();
  const method = new Method('foo');
  method.setReturnReference(true);

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints method with reference parameters', () => {
  const printer = new Printer();
  const method = new Method('foo');
  const param = method.addParameter('param');
  param.setReference(true);

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});

test('prints method in interface context', () => {
  const printer = new Printer();
  const method = new Method('foo');
  method.setReturnType('string');
  method.addParameter('param');

  const output = printer.printMethod(method, null, true);

  expect(output).toMatchSnapshot();
});

test('prints complex method', () => {
  const printer = new Printer();
  const method = new Method('complexMethod');

  // Set modifiers
  method.setPublic();
  method.setStatic();

  // Add doc comment
  method.addComment('Complex method with multiple features');
  method.addComment('@param string $param1 First parameter');
  method.addComment('@param int $param2 Second parameter');
  method.addComment('@return array<string>');

  // Add parameters
  const param1 = method.addParameter('param1');
  param1.setType('string');
  param1.setNullable(true);

  const param2 = new PromotedParameter('param2');
  param2.setType('int');
  param2.setReadOnly(true);
  method.setParameters([param1, param2]);

  // Set return type
  method.setReturnType('array');

  // Add attribute
  method.addAttribute('Route', { path: '/complex' });

  const output = printer.printMethod(method);

  expect(output).toMatchSnapshot();
});
