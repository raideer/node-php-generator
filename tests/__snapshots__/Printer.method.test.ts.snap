// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prints abstract method 1`] = `
"abstract function foo();
"
`;

exports[`prints basic method 1`] = `
"function foo()
{
}
"
`;

exports[`prints complex method 1`] = `
"/**
 * Complex method with multiple features
 * @param string $param1 First parameter
 * @param int $param2 Second parameter
 * @return array<string>
 */
#[Route(path: '/complex')]
public static function complexMethod(
    ?string $param1,
    public readonly int $param2,
): array
{
}
"
`;

exports[`prints final method 1`] = `
"final function foo()
{
}
"
`;

exports[`prints method in interface context 1`] = `
"function foo($param): string;
"
`;

exports[`prints method with attributes 1`] = `
"#[Route(path: '/foo')]
function foo()
{
}
"
`;

exports[`prints method with doc comment 1`] = `
"/**
 * This is a method
 * @param string $param Description
 * @return void
 */
function foo()
{
}
"
`;

exports[`prints method with nullable return type 1`] = `
"function foo(): ?string
{
}
"
`;

exports[`prints method with parameters 1`] = `
"function foo($param1 = 'default1', $param2 = 'default2')
{
}
"
`;

exports[`prints method with promoted parameters 1`] = `
"function foo(
    public $param = 'default',
) {
}
"
`;

exports[`prints method with reference parameters 1`] = `
"function foo(&$param)
{
}
"
`;

exports[`prints method with reference return 1`] = `
"function &foo()
{
}
"
`;

exports[`prints method with return type 1`] = `
"function foo(): string
{
}
"
`;

exports[`prints method with typed parameters 1`] = `
"function foo(?string $param)
{
}
"
`;

exports[`prints private method 1`] = `
"private function foo()
{
}
"
`;

exports[`prints protected method 1`] = `
"protected function foo()
{
}
"
`;

exports[`prints public method 1`] = `
"public function foo()
{
}
"
`;

exports[`prints static method 1`] = `
"static function foo()
{
}
"
`;
